tuning.py

```diff
@@ -0,0 +1,308 @@
+# coding=utf-8
+# Copyright 2021 The Google Research Authors.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Landscape tuning."""
+
+import itertools
+from pprint import pprint
+from typing import Optional, Tuple, Dict
+
+import pandas as pd
+import numpy as np
+from scipy.special import comb
+
+import potts_model
+import sampling
+import experiment
+import utils
+
+
+def get_fraction_adaptive_singles(landscape: potts_model.PottsModel) -> float:
+    """Returns the fraction of singles that are adaptive."""
+    all_singles = sampling.get_all_single_mutants(landscape.wildtype_sequence, landscape.vocab_size)
+    wt_fitness = landscape.evaluate(landscape.wildtype_sequence).item()
+    fraction_adaptive = (landscape.evaluate(all_singles) >= wt_fitness).sum() / all_singles.shape[0]
+    return fraction_adaptive
+
+
+def get_doubles_df(landscape: potts_model.PottsModel, threshold: float, adaptive: bool) -> pd.DataFrame:
+    """Returns a DataFrame of all combinations of singles that are above/below a given fitness threshold.
```

Comment:
I assume you mean pairwise combination here? Should make that clear.

Reply:
Also, the way it's worded makes it sound like the conditioning is on the doubles, not the singles.

Reply:
Agreed, I'll elaborate in the docstring